sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/PlanCustomizer.java

```diff
@@ -4,5 +4,13 @@
  * This interface allows an opportunity for plans to be modified before execution.
  */
 public interface PlanCustomizer {
+
+    /**
+     * Takes a reference to a plan, returning a reference to a modified plan derived from the original.
+     *
+     * Note: ALL plans (including the uninstall deploy plan) are run through the customizer if one is declared.
+     * @param plan
+     * @return
+     */
```

Comment:
Would it make sense to just add an explicit 'Mode' enum? Modes could be DEPLOY or UNINSTALL for now, possibly another someday for e.g. UPDATE. Feels like most customizers will only be applicable to a subset of modes. Also in a module world I'd expect SchedulerConfig to be inaccessible accessible...

Could either update the existing function to have this param, or add a separate default/overrideable updatePlan(Plan, Mode) function which by default only calls the currently-defined updatePlan(Plan) function if not uninstalling (retaining cur semantics for updatePlan(Plan), if it makes sense to do that).

Reply:
Gabriel and I had discussed out of band. I think the three options are:

0. The plan customizer will run updatePlan() against any plan, regardless of whether uninstall is happening or not. They'd then need to always gate the `deploy` plan with the `SchedulerConfig.isUninstallEnabled` flag.
1. Add a second method to the interface (e.g. updateUninstallPlan)
2. Add a `Mode` argument to the method.

I'm okay with any of these options, my only thing is making sure the user is not confused and knows what they need to do.

I hadn't considered how this would work under modularization. We'd definitely want the user to be able to see that uninstall is active, in their module. What you suggest, seems like a reasonable way to accomplish that.

Reply:
Option 1 feels like it'd be the best option. We'd probably want to also have both the functions in the `PlanCustomizer` interface then be `default` functions with noop behavior.

This would specifically make it a bit easier if we wanted to add an additional mode later. With default functions that do nothing, developers wouldn't technically need to do anything in order to be compatible, whereas with an enum they'd need to update their switch statement or similar when bumping the SDK.

Reply:
Oh, that is a good point.

I am going with option 1, as we need to practice being good citizens.